File: packages/driver/src/mlx-ml/process/prompt-builder.ts
/**
 * chat template 非対応モデル向けのプロンプト組立（Python generate_merged_prompt の TS 移植）
 */

import type { InferenceCapabilities, InferenceMessage } from '../../local-inference/protocol.js';
import type { SpecialToken, SpecialTokenPair } from '../../formatter/types.js';

type SpecialTokensMap = InferenceCapabilities['special_tokens'];

function messageContentToString(content: InferenceMessage['content']): string {
  if (typeof content === 'string') {
    return content;
  }
  return content
    .filter((part): part is { type: 'text'; text: string } => part.type === 'text')
    .map((part) => part.text)
    .join('');
}

function isTokenPair(token: SpecialToken | SpecialTokenPair | undefined): token is SpecialTokenPair {
  return !!token && typeof token === 'object' && 'start' in token;
}

/**
 * apply_chat_template が使えない場合に special_tokens からプロンプトを組み立てる
 */
export function generateMergedPrompt(
  messages: InferenceMessage[],
  specialTokens: SpecialTokensMap,
): string {
  const promptParts: string[] = [];

  for (const msg of messages) {
    const role = msg.role;
    const roleUpper = role.toUpperCase();
    const content = messageContentToString(msg.content).trim();

    const roleToken = specialTokens[role];
    if (isTokenPair(roleToken)) {
      promptParts.push(
        roleToken.start.text,
        content,
        roleToken.end.text,
        '',
      );
      continue;
    }

    let blockToken: SpecialTokenPair | undefined;
    for (const candidate of ['block', 'context', 'quote', 'section'] as const) {
      const token = specialTokens[candidate];
      if (isTokenPair(token)) {
        blockToken = token;
        break;
      }
    }

    if (blockToken) {
      promptParts.push(
        `${blockToken.start.text}${roleUpper}:\n${content}`,
        blockToken.end.text,
        '',
      );
    } else {
      promptParts.push(
        `<!-- begin of ${roleUpper} -->`,
        content,
        `<!-- end of ${roleUpper} -->`,
        '',
      );
    }
  }

  if (promptParts.length === 0) {
    return '';
  }
  promptParts.pop();
  return promptParts.join('\n');
}

File: packages/driver/src/pytorch/templates/cuda/utils/prompt_builder.py
"""プロンプト生成ユーティリティ"""


def supports_chat_template(tokenizer) -> bool:
    return (hasattr(tokenizer, 'apply_chat_template') and
            hasattr(tokenizer, 'chat_template') and
            tokenizer.chat_template is not None)


def generate_merged_prompt(messages, capabilities):
    """apply_chat_templateがない場合のプロンプト生成"""
    prompt_parts = []
    special_tokens = capabilities.get('special_tokens', {})

    for msg in messages:
        role = msg['role']
        role_upper = role.upper()

        role_token = special_tokens.get(role)

        if role_token and isinstance(role_token, dict) and 'start' in role_token:
            start_token = role_token['start']['text']
            end_token = role_token['end']['text']
            prompt_parts.extend([
                start_token,
                msg['content'].strip(),
                end_token,
                ''
            ])
        else:
            block_token = None
            for candidate in ['block', 'context', 'quote', 'section']:
                token = special_tokens.get(candidate)
                if token and isinstance(token, dict) and 'start' in token:
                    block_token = token
                    break

            if block_token:
                start_token = block_token['start']['text']
                end_token = block_token['end']['text']
                prompt_parts.extend([
                    f'{start_token}{role_upper}:\n{msg["content"].strip()}',
                    end_token,
                    ''
                ])
            else:
                prompt_parts.extend([
                    f'<!-- begin of {role_upper} -->',
                    msg['content'].strip(),
                    f'<!-- end of {role_upper} -->',
                    ''
                ])

    return '\n'.join(prompt_parts[:-1])
